File: Poker-Basic/src/main/java/com/pokermon/AdventureMode.java
package com.pokermon;

import java.util.Random;
import java.util.Scanner;

/**
 * Adventure Mode implementation - Monster battles using poker combat.
 * Players battle monsters whose health equals their chip count.
 * Poker hand strength determines battle outcomes.
 * 
 * @author Carl Nelson (@Gameaday)
 * @version 1.0.0 - Beta Implementation
 */
public class AdventureMode {
    
    private static final Scanner scanner = new Scanner(System.in);
    private static final Random random = new Random();
    
    private final String playerName;
    private final int initialChips;
    private MonsterCollection playerCollection;
    private int currentLevel;
    private int monstersDefeated;
    
    /**
     * Creates a new Adventure Mode game session.
     */
    public AdventureMode(String playerName, int initialChips) {
        this.playerName = playerName;
        this.initialChips = initialChips;
        this.playerCollection = new MonsterCollection();
        this.currentLevel = 1;
        this.monstersDefeated = 0;
    }
    
    /**
     * Starts the Adventure Mode gameplay loop.
     */
    public void startAdventure() {
        System.out.println("🏔️ " + playerName + "'s Adventure Begins! 🏔️");
        System.out.println("You start with " + initialChips + " chips to use in battle.");
        System.out.println();
        
        boolean continueAdventure = true;
        int playerChips = initialChips;
        
        while (continueAdventure && playerChips > 0) {
            Monster enemy = generateEnemyMonster();
            System.out.println("💀 A wild " + enemy.getName() + " appears!");
            System.out.println("   Rarity: " + enemy.getRarity().getDisplayName());
            System.out.println("   Health: " + enemy.getEffectiveHealth() + " HP");
            System.out.println("   Your chips: " + playerChips);
            System.out.println();
            
            BattleResult result = battleMonster(enemy, playerChips);
            
            if (result.victory) {
                System.out.println("🎉 Victory! You defeated the " + enemy.getName() + "!");
                
                // Calculate rewards
                int chipReward = calculateChipReward(enemy);
                playerChips += chipReward;
                monstersDefeated++;
                
                System.out.println("   Chip reward: +" + chipReward + " chips");
                System.out.println("   Total chips: " + playerChips);
                
                // Chance to capture the monster
                if (attemptCapture(enemy, result.handStrength)) {
                    playerCollection.addMonster(enemy);
                    System.out.println("   🎁 Bonus: " + enemy.getName() + " joined your collection!");
                }
                
                System.out.println();
                currentLevel++;
                
            } else {
                System.out.println("💀 Defeat! The " + enemy.getName() + " was too strong.");
                int chipsLost = Math.min(playerChips / 4, 100); // Lose 25% or max 100
                playerChips -= chipsLost;
                System.out.println("   You lost " + chipsLost + " chips in the retreat.");
                System.out.println("   Remaining chips: " + playerChips);
                System.out.println();
            }
            
            if (playerChips <= 0) {
                System.out.println("💀 Game Over! You've run out of chips for battle.");
                break;
            }
            
            continueAdventure = promptContinue();
        }
        
        showAdventureStats();
    }
    
    /**
     * Generates an enemy monster based on current level.
     */
    private Monster generateEnemyMonster() {
        Monster.Rarity rarity;
        String name;
        int baseHealth;
        
        // Determine rarity based on level (higher level = rarer monsters)
        int rarityRoll = random.nextInt(100);
        if (currentLevel >= 10 && rarityRoll < 5) {
            rarity = Monster.Rarity.LEGENDARY;
            name = "Ancient Dragon";
            baseHealth = 800;
        } else if (currentLevel >= 7 && rarityRoll < 15) {
            rarity = Monster.Rarity.EPIC;
            name = "Shadow Beast";
            baseHealth = 500;
        } else if (currentLevel >= 4 && rarityRoll < 30) {
            rarity = Monster.Rarity.RARE;
            name = "Crystal Golem";
            baseHealth = 300;
        } else if (currentLevel >= 2 && rarityRoll < 60) {
            rarity = Monster.Rarity.UNCOMMON;
            name = "Forest Guardian";
            baseHealth = 150;
        } else {
            rarity = Monster.Rarity.COMMON;
            name = "Wild Slime";
            baseHealth = 100;
        }
        
        // Scale health with level
        int scaledHealth = baseHealth + (currentLevel - 1) * 50;
        
        return new Monster(
            name + " Lv." + currentLevel,
            rarity,
            scaledHealth, // baseHealth 
            Monster.EffectType.CHIP_BONUS,
            20, // effectPower
            "A wild monster encountered in adventure mode"
        );
    }
    
    /**
     * Conducts a poker battle between player and monster.
     */
    private BattleResult battleMonster(Monster enemy, int playerChips) {
        System.out.println("⚔️ POKER BATTLE COMMENCES! ⚔️");
        System.out.println("Draw your hand and prepare for battle!");
        System.out.println();
        
        // Create a simple poker hand for the player
        Player tempPlayer = new Player();
        tempPlayer.setName(playerName);
        tempPlayer.setChipsCurrent(playerChips);
        int[] deck = Main.setDeck();
        
        // Initialize player with a poker hand
        int[] hand = Main.newHand(deck);
        tempPlayer.updateHand(hand);
        tempPlayer.convertHand();
        tempPlayer.calculateHandValue();
        
        // Display player's hand
        String[] playerHand = tempPlayer.getConvertedHand();
        System.out.println("Your battle hand:");
        for (int i = 0; i < playerHand.length; i++) {
            System.out.println("  " + (i + 1) + ": " + playerHand[i]);
        }
        System.out.println();
        
        int handValue = tempPlayer.getHandValue();
        System.out.println("Hand strength: " + handValue + " (" + getHandDescription(handValue) + ")");
        
        // Calculate battle outcome
        int damageDealt = calculateDamage(handValue, enemy);
        boolean victory = damageDealt >= enemy.getEffectiveHealth();
        
        System.out.println("⚡ You deal " + damageDealt + " damage!");
        
        if (victory) {
            System.out.println("💥 Critical hit! The monster is defeated!");
        } else {
            System.out.println("💢 The monster survives with " + (enemy.getEffectiveHealth() - damageDealt) + " HP remaining.");
        }
        System.out.println();
        
        return new BattleResult(victory, handValue, damageDealt);
    }
    
    /**
     * Calculates damage based on poker hand strength.
     */
    private int calculateDamage(int handValue, Monster enemy) {
        // Base damage scales with hand strength
        int baseDamage = handValue * 10;
        
        // Add randomness (80-120% of base damage)
        double multiplier = 0.8 + (random.nextDouble() * 0.4);
        int finalDamage = (int) (baseDamage * multiplier);
        
        // Rarity affects enemy toughness
        double rarityDefense = 1.0 / enemy.getRarity().getPowerMultiplier();
        finalDamage = (int) (finalDamage * rarityDefense);
        
        return Math.max(finalDamage, 1); // Minimum 1 damage
    }
    
    /**
     * Calculates chip rewards for defeating a monster.
     */
    private int calculateChipReward(Monster enemy) {
        int baseReward = enemy.getEffectiveHealth() / 2; // Half the monster's health
        double rarityBonus = enemy.getRarity().getPowerMultiplier();
        return (int) (baseReward * rarityBonus);
    }
    
    /**
     * Attempts to capture a defeated monster.
     */
    private boolean attemptCapture(Monster enemy, int handStrength) {
        // Higher hand strength = better capture chance
        double captureChance = Math.min(0.8, 0.1 + (handStrength * 0.02));
        
        // Rarer monsters are harder to capture
        captureChance /= enemy.getRarity().getPowerMultiplier();
        
        return random.nextDouble() < captureChance;
    }
    
    /**
     * Gets a description of the poker hand strength.
     */
    private String getHandDescription(int handValue) {
        if (handValue >= 80) return "Royal Flush Territory";
        if (handValue >= 60) return "Four of a Kind/Straight Flush";
        if (handValue >= 40) return "Full House/Flush";
        if (handValue >= 30) return "Straight/Three of a Kind";
        if (handValue >= 20) return "Two Pair";
        if (handValue >= 10) return "One Pair";
        return "High Card";
    }
    
    /**
     * Prompts player to continue the adventure.
     */
    private boolean promptContinue() {
        System.out.print("Continue your adventure? (y/n) [y]: ");
        String input = scanner.nextLine().trim().toLowerCase();
        return input.isEmpty() || input.equals("y") || input.equals("yes");
    }
    
    /**
     * Displays final adventure statistics.
     */
    private void showAdventureStats() {
        System.out.println("🏆 ADVENTURE COMPLETE! 🏆");
        System.out.println("=".repeat(40));
        System.out.println("Monsters defeated: " + monstersDefeated);
        System.out.println("Highest level reached: " + currentLevel);
        System.out.println("Monsters captured: " + playerCollection.getMonsterCount());
        System.out.println();
        
        if (playerCollection.getMonsterCount() > 0) {
            System.out.println("Your Monster Collection:");
            // Note: MonsterCollection would need a method to list monsters
            // For now, just show count
            System.out.println("  Total monsters: " + playerCollection.getMonsterCount());
        }
        
        System.out.println("Thank you for playing Adventure Mode!");
    }
    
    /**
     * Result of a battle encounter.
     */
    private static class BattleResult {
        final boolean victory;
        final int handStrength;
        final int damageDealt;
        
        BattleResult(boolean victory, int handStrength, int damageDealt) {
            this.victory = victory;
            this.handStrength = handStrength;
            this.damageDealt = damageDealt;
        }
    }
}

File: Poker-Basic/src/main/kotlin/com/pokermon/AdventureMode.kt
package com.pokermon

import kotlin.math.min
import kotlin.math.max
import kotlin.random.Random

/**
 * Adventure Mode implementation - Monster battles using poker combat.
 * Players battle monsters whose health equals their chip count.
 * Poker hand strength determines battle outcomes.
 * 
 * @author Carl Nelson (@Gameaday)
 * @version 1.0.0 - Beta Implementation (Kotlin-native)
 */
class AdventureMode(
    private val playerName: String,
    private val initialChips: Int
) {
    private val playerCollection = MonsterCollection()
    private var currentLevel = 1
    private var monstersDefeated = 0
    
    companion object {
        private val random = Random.Default
    }
    
    /**
     * Starts the Adventure Mode gameplay loop.
     */
    fun startAdventure() {
        println("🏔️ ${playerName}'s Adventure Begins! 🏔️")
        println("You start with $initialChips chips to use in battle.")
        println()
        
        var continueAdventure = true
        var playerChips = initialChips
        
        while (continueAdventure && playerChips > 0) {
            val enemy = generateEnemyMonster()
            println("💀 A wild ${enemy.name} appears!")
            println("   Rarity: ${enemy.rarity.displayName}")
            println("   Health: ${enemy.effectiveHealth} HP")
            println("   Your chips: $playerChips")
            println()
            
            val result = battleMonster(enemy, playerChips)
            
            if (result.victory) {
                println("🎉 Victory! You defeated the ${enemy.name}!")
                
                // Calculate rewards
                val chipReward = calculateChipReward(enemy)
                playerChips += chipReward
                monstersDefeated++
                
                println("   Chip reward: +$chipReward chips")
                println("   Total chips: $playerChips")
                
                // Chance to capture the monster
                if (attemptCapture(enemy, result.handStrength)) {
                    playerCollection.addMonster(enemy)
                    println("   🎁 Bonus: ${enemy.name} joined your collection!")
                }
                
                println()
                currentLevel++
                
            } else {
                println("💀 Defeat! The ${enemy.name} was too strong.")
                val chipsLost = min(playerChips / 4, 100) // Lose 25% or max 100
                playerChips -= chipsLost
                println("   You lost $chipsLost chips in the retreat.")
                println("   Remaining chips: $playerChips")
                println()
            }
            
            if (playerChips <= 0) {
                println("💀 Game Over! You've run out of chips for battle.")
                break
            }
            
            continueAdventure = promptContinue()
        }
        
        showAdventureStats()
    }
    
    /**
     * Generates an enemy monster based on current level.
     */
    private fun generateEnemyMonster(): Monster {
        val (rarity, name, baseHealth) = when {
            currentLevel >= 10 && random.nextInt(100) < 5 -> {
                Triple(Monster.Rarity.LEGENDARY, "Ancient Dragon", 800)
            }
            currentLevel >= 7 && random.nextInt(100) < 15 -> {
                Triple(Monster.Rarity.EPIC, "Shadow Beast", 500)
            }
            currentLevel >= 4 && random.nextInt(100) < 30 -> {
                Triple(Monster.Rarity.RARE, "Crystal Golem", 300)
            }
            currentLevel >= 2 && random.nextInt(100) < 60 -> {
                Triple(Monster.Rarity.UNCOMMON, "Forest Guardian", 150)
            }
            else -> {
                Triple(Monster.Rarity.COMMON, "Wild Slime", 100)
            }
        }
        
        // Scale health with level
        val scaledHealth = baseHealth + (currentLevel - 1) * 50
        
        return Monster(
            name = "$name Lv.$currentLevel",
            rarity = rarity,
            baseHealth = scaledHealth,
            effectType = Monster.EffectType.CHIP_BONUS,
            effectPower = 20,
            description = "A wild monster encountered in adventure mode"
        )
    }
    
    /**
     * Conducts a poker battle between player and monster.
     */
    private fun battleMonster(enemy: Monster, playerChips: Int): BattleResult {
        println("⚔️ POKER BATTLE COMMENCES! ⚔️")
        println("Draw your hand and prepare for battle!")
        println()
        
        // Create a simple poker hand for the player using modern game engine
        val gameEngine = GameEngine(
            gameConfig = Game(
                gameMode = GameMode.ADVENTURE,
                startingChips = playerChips
            )
        )
        
        // Initialize the game with the player
        gameEngine.initializeGame(arrayOf(playerName))
        val hand = gameEngine.players!![0].hand
        
        // Evaluate hand using HandEvaluator for consistent scoring
        val handResult = HandEvaluator.evaluateHand(hand)
        val handValue = handResult.score
        
        // Display player's hand using CardUtils
        println("Your battle hand:")
        hand.forEachIndexed { index, card ->
            println("  ${index + 1}: ${CardUtils.cardName(card)}")
        }
        println()
        
        println("Hand strength: $handValue (${handResult.description})")
        
        // Calculate battle outcome
        val damageDealt = calculateDamage(handValue, enemy)
        val victory = damageDealt >= enemy.effectiveHealth
        
        println("⚡ You deal $damageDealt damage!")
        
        if (victory) {
            println("💥 Critical hit! The monster is defeated!")
        } else {
            println("💢 The monster survives with ${enemy.effectiveHealth - damageDealt} HP remaining.")
        }
        println()
        
        return BattleResult(victory, handValue, damageDealt)
    }
    
    /**
     * Calculates damage based on poker hand strength.
     */
    private fun calculateDamage(handValue: Int, enemy: Monster): Int {
        // Base damage scales with hand strength
        val baseDamage = handValue * 10
        
        // Add randomness (80-120% of base damage)
        val multiplier = 0.8 + (random.nextDouble() * 0.4)
        var finalDamage = (baseDamage * multiplier).toInt()
        
        // Rarity affects enemy toughness
        val rarityDefense = 1.0 / enemy.rarity.powerMultiplier
        finalDamage = (finalDamage * rarityDefense).toInt()
        
        return max(finalDamage, 1) // Minimum 1 damage
    }
    
    /**
     * Calculates chip rewards for defeating a monster.
     */
    private fun calculateChipReward(enemy: Monster): Int {
        val baseReward = enemy.effectiveHealth / 2 // Half the monster's health
        val rarityBonus = enemy.rarity.powerMultiplier
        return (baseReward * rarityBonus).toInt()
    }
    
    /**
     * Attempts to capture a defeated monster.
     */
    private fun attemptCapture(enemy: Monster, handStrength: Int): Boolean {
        // Higher hand strength = better capture chance
        var captureChance = min(0.8, 0.1 + (handStrength * 0.02))
        
        // Rarer monsters are harder to capture
        captureChance /= enemy.rarity.powerMultiplier
        
        return random.nextDouble() < captureChance
    }
    
    /**
     * Prompts player to continue the adventure.
     */
    private fun promptContinue(): Boolean {
        print("Continue your adventure? (y/n) [y]: ")
        val input = readlnOrNull()?.trim()?.lowercase() ?: ""
        return input.isEmpty() || input == "y" || input == "yes"
    }
    
    /**
     * Displays final adventure statistics.
     */
    private fun showAdventureStats() {
        println("🏆 ADVENTURE COMPLETE! 🏆")
        println("=".repeat(40))
        println("Monsters defeated: $monstersDefeated")
        println("Highest level reached: $currentLevel")
        println("Monsters captured: ${playerCollection.getMonsterCount()}")
        println()
        
        if (playerCollection.getMonsterCount() > 0) {
            println("Your Monster Collection:")
            println("  Total monsters: ${playerCollection.getMonsterCount()}")
            playerCollection.getOwnedMonsters().forEach { monster ->
                println("  - ${monster.name} (${monster.rarity.displayName})")
            }
        }
        
        println("Thank you for playing Adventure Mode!")
    }
    
    /**
     * Result of a battle encounter.
     */
    private data class BattleResult(
        val victory: Boolean,
        val handStrength: Int,
        val damageDealt: Int
    )
}